tpm_futurepcr/LogEvent.py

```diff
@@ -0,0 +1,141 @@
+                    logger.verbose("  * %-20s %-20s %s", type_name, subtype_name, file_path)
+        elif self.type in {TpmEventType.EFI_VARIABLE_AUTHORITY,
+                            TpmEventType.EFI_VARIABLE_BOOT,
+                            TpmEventType.EFI_VARIABLE_DRIVER_CONFIG}:
+            if logger.level == logging.DEBUG:
+                for i in hexdump(self.data, 64):
+                    logger.debug(i)
+                ed = self._parse_efi_variable_event(self.data)
+                logger.debug(pformat(ed))
+            else:
+                ed = self._parse_efi_variable_event(self.data)
+                logger.verbose("Variable: %r {%s}", ed["unicode_name"], ed["variable_name_uuid"])
+        else:
+            for i in hexdump(self.data, 64):
+                logger.debug(i)
+
+    def __post_init__(self, binary_reader):
+        self.pcr_idx = binary_reader.read(READFMT.U32)
+        self.type = TpmEventType(binary_reader.read(READFMT.U32))
+        self.pcr_extend_values = dict()
+        # same across both formats
+        if LogEvent.tpm_version == 1:
+            # section 5.1, SHA1 Event Log Entry Format
+            self.pcr_extend_values[TpmAlgorithm.SHA1] = binary_reader.read(20)
+        elif LogEvent.tpm_version == 2:
+            # section 5.2, Crypto Agile Log Entry Format
+            pcr_count = binary_reader.read(READFMT.U32)
+            for i in range(pcr_count):
+                # Spec says it should be safe to just iter over hdr[digest_sizes],
+                # as all entries must have the same algorithms in the same order,
+                # but it does recommend alg_id lookup as the preferred method.
+                _alg = binary_reader.read(READFMT.U16)
+                alg_id = TpmAlgorithm(_alg)
+                self.pcr_extend_values[alg_id] = binary_reader.read(self.tcg_hdr["digest_sizes_dict"][alg_id])
+
+        # same across both formats
+        event_size = binary_reader.read(READFMT.U32)
+        if self.type == TpmEventType.EFI_BOOT_SERVICES_APPLICATION:
+            image_location = binary_reader.read(READFMT.PTR)     # EFI_PHYSICAL_ADDRESS (pointer)
+            image_length = binary_reader.read(READFMT.SIZE)      # UINTN (u64/u32 depending on arch)
+            image_lt_address = binary_reader.read(READFMT.SIZE)  # UINTN
+            device_path_len = binary_reader.read(READFMT.SIZE)   # UINTN
+            device_path_vec = parse_efi_device_path(binary_reader.read(device_path_len))
+            self.data = EFIBSAData(image_location, image_length, image_lt_address, device_path_vec)
+        else:
+            self.data = binary_reader.read(event_size)
+
+            if LogEvent.tpm_version == 1 and self.pcr_idx == 0 and \
+               self.type == TpmEventType.NO_ACTION and self.data[:15] == b"Spec ID Event03":
+                LogEvent.tpm_version = 2
+                LogEvent.tcg_hdr = self._parse_efi_tcg2_header_event(self.data)
```

Comment:
While LogEvent in itself is an improvement over plain dicts, please don't introduce this kind of global state – it's not just about tests having to reset it, but is bad design in general.

Instead of trying to do everything in the dataclass constructor, might be better to let `enum_log_entries` keep track of file-level state (which is the purpose of that function, after all) and pass both `tpm_version` and `tcg_hdr` to the LogEvent parser as needed.

Reply:
yes, I know. I did not want to move things around more than required (otherwise even more changes come in the same PR). I am fixing it and adding another commit to this one.